File: 28_07_2024_Second_Minimum_Time_to_Reach_Destination.java
class Solution {
    public int secondMinimum(int n, int[][] edges, int time, int change) {
        
        Map<Integer, List<Integer>> g = new HashMap<>();
        for (int[] e : edges) {
            int u = e[0], v = e[1];
            g.computeIfAbsent(u, k -> new ArrayList<>()).add(v);
            g.computeIfAbsent(v, k -> new ArrayList<>()).add(u);
        }

        PriorityQueue<int[]> q = new PriorityQueue<>(Comparator.comparingInt(a -> a[0]));
        q.offer(new int[]{0, 1}); 

        int[] uniqueVisit = new int[n + 1];  
        int[] dis = new int[n + 1];  
        Arrays.fill(dis, -1);
        
        while (!q.isEmpty()) {
            int[] current = q.poll();
            int t = current[0], node = current[1];
            
            if (dis[node] == t || uniqueVisit[node] >= 2) {
                continue;
            }
            
            uniqueVisit[node]++;
            dis[node] = t;
            
            if (node == n && uniqueVisit[node] == 2) {
                return dis[node];
            }
            
            if ((t / change) % 2 != 0) {
                t = (t / change + 1) * change;
            }
            
            for (int nei : g.getOrDefault(node, new ArrayList<>())) {
                q.offer(new int[]{t + time, nei});
            }
        }
        
        return -1;
    }
}

File: 28_07_2024_Second_Minimum_Time_to_Reach_Destination.py
from collections import defaultdict
import heapq

class Solution:
    def secondMinimum(self, n: int, edges: list[list[int]], time: int, change: int) -> int:
        g = defaultdict(list)
        for e in edges:
            u, v = e[0], e[1]
            g[u].append(v)
            g[v].append(u)

        q = []
        heapq.heappush(q, (0, 1))

        uniqueVisit = [0] * (n + 1)
        dis = [-1] * (n + 1)

        while q:
            t, node = heapq.heappop(q)

            if dis[node] == t or uniqueVisit[node] >= 2:
                continue

            uniqueVisit[node] += 1
            dis[node] = t

            if node == n and uniqueVisit[node] == 2:
                return dis[node]

            if (t // change) % 2 != 0:
                t = (t // change + 1) * change

            for nei in g[node]:
                heapq.heappush(q, (t + time, nei))

        return -1

